tests/DependencyInjection/Compiler/RegisterMenusPassTest.php

```diff
@@ -21,93 +22,93 @@ class RegisterMenusPassTest extends TestCase
     {
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu_builder', true)->willReturn([
-            'id1' => [['alias' => 'foo', 'method' => 'fooMenu'], ['alias' => 'bar', 'method' => 'bar']],
-            'id2' => [['alias' => 'foo', 'method' => 'fooBar'], ['alias' => 'baz', 'method' => 'bar']],
-        ]);
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu', true)->willReturn(['id3' => [['alias' => 'test']]]);
-
-        $menus = [
-            'foo' => [new ServiceClosureArgument(new Reference('id2')), 'fooBar'],
-            'bar' => [new ServiceClosureArgument(new Reference('id1')), 'bar'],
-            'baz' => [new ServiceClosureArgument(new Reference('id2')), 'bar'],
-            'test' => new ServiceClosureArgument(new Reference('id3')),
-        ];
-        $this->definition->replaceArgument(0, $menus)->shouldBeCalled();
-
-        $this->pass->process($this->containerBuilder->reveal());
+        $this->pass->process($this->containerBuilder);
+
+        $menuBuilders = $this->containerBuilder->getDefinition('knp_menu.menu_provider.lazy')->getArgument(0);
+
+        $this->assertFalse($this->containerBuilder->hasDefinition('knp_menu.menu_provider.builder_service'));
+        $this->assertArrayHasKey('foo', $menuBuilders);
+        $this->assertArrayHasKey('bar', $menuBuilders);
+        $this->assertArrayHasKey('baz', $menuBuilders);
+        $this->assertIsArray($menuBuilders['bar']);
+        $this->assertIsArray($menuBuilders['foo']);
+        $this->assertSame('id', (string) $menuBuilders['foo'][0]->getValues()[0]);
+        $this->assertSame('id', (string) $menuBuilders['bar'][0]->getValues()[0]);
+        $this->assertSame('fooMenu', $menuBuilders['foo'][1]);
+        $this->assertSame('bar', $menuBuilders['bar'][1]);
     }
 
     public function testMenuWinsOverBuilder(): void
     {
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu_builder', true)->willReturn([
-            'id1' => [['alias' => 'foo', 'method' => 'fooMenu'], ['alias' => 'bar', 'method' => 'bar']],
-        ]);
-        $this->containerBuilder->findTaggedServiceIds('knp_menu.menu', true)->willReturn(['id3' => [['alias' => 'foo']]]);
-
-        $menus = [
-            'foo' => new ServiceClosureArgument(new Reference('id3')),
-            'bar' => [new ServiceClosureArgument(new Reference('id1')), 'bar'],
-        ];
-        $this->definition->replaceArgument(0, $menus)->shouldBeCalled();
-
-        $this->pass->process($this->containerBuilder->reveal());
+        $this->containerBuilder->getDefinition('menu_id')
+            ->setTags(['knp_menu.menu' => [['alias' => 'foo']]]);
+
+        $this->pass->process($this->containerBuilder);
+
+        $menuBuilders = $this->containerBuilder->getDefinition('knp_menu.menu_provider.lazy')->getArgument(0);
+
+        $this->assertArrayHasKey('foo', $menuBuilders);
+        $this->assertIsNotArray($menuBuilders['foo']);
```

Comment:
not being an array does not guarantee it is a ServiceClosureArgument

Reply:
Fixed by checking with the complete array of expected results.